File: DSA/1.2_Arrays/2.1_Two_Pointers/is_subsequence.js
/*
i
ahbgdc

abc
j
*/

function isSubsequence(s, t) { // non-congruent

  let count = 0
  let j = 0
  for (let i = 0; i < t.length; i++) {
    if (t[i] === s[j]) {
      count++
      j++
    }
  }

  if (count === s.length) return true
  else return false

  // or can do return (j === s.length) ? true : false
  // but (j === s.length) ? return true : return false is wrong
}

console.log(isSubsequence('abc', 'ahbgdc')); // true
isSubsequence('sing', 'sting'); // true
isSubsequence('abc', 'abracadabra'); // true
console.log(isSubsequence('abc', 'acb')); // false (order matters)

/*
Logic is keep looping on string and after each char is matched on the substring, increment count
when count === length of the substring, then a subsequence exists
*/

File: DSA/1.2_Arrays/2.1_Two_Pointers/is_subsequence.py
def isSubsequence(needle, haystack):

    j = 0
    count = 0

    for i in range(len(haystack)):

        if haystack[i] == needle[j]:
            j += 1
            count += 1

        if count == len(needle): return True

    return False


print(isSubsequence('abc', 'ahbgdc')); # true
print(isSubsequence('sing', 'sting')); # true
print(isSubsequence('abc', 'abracadabra')); # true
print(isSubsequence('abc', 'acb')); # false (order matters)